### PR TITLE
fix: get token metadata from onchain

### LINKED ISSUE #10875: Evm Gallery Item missing animation
### What happened?

https://github.com/user-attachments/assets/debacecf-adab-41ca-b4c5-926e786064ad

### LINKED ISSUE #10926: Weird BUG: Fullscreen showing different NFT
So this happened. I opened [this](https://kodadot.xyz/ahp/gallery/99-169) generative NFT, clicked on fullscreen and it opened totally different NFT. 

https://github.com/user-attachments/assets/6ba6eaf9-03d4-4d4e-9d9d-6a02e4c2c75b

### LINKED ISSUE #10927: Animated collection is static
Collection: https://kodadot.xyz/base/drops/rvn3

Whole collection should be animated and its static

https://github.com/user-attachments/assets/71c6a094-11eb-4f3f-8372-64f2415ebb0d

### LINKED ISSUE #11058: Open in new tab: opens different NFT
Link: https://kodadot.xyz/ahp/gallery/241-4227615263
Polkadot drop - when you open this NFT in a new tab, a different NFT opens. We had this issue before, but I thought it was fixed. Pls can you guys fix it? The artist wants to download all the NFTs from that collection in high quality!

https://github.com/user-attachments/assets/4b19c5d4-48fa-44fb-8f82-4338dd7cbb60

### PR DESCRIPTION
**Pull request by preschian**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring

## Context

- [x] Closes #10926
- [x] Closes #10927
- [x] Closes #10875
- [x] Closes #11058
- [x] Requires deployment https://github.com/kodadot/private-workers/pull/216

tldr; here: https://github.com/kodadot/nft-gallery/pull/10934#issuecomment-2333386275

#### Did your issue had any of the "$" label on it?

- [x] My DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer/?target=1xjvRADwdJcnmUCLWerEHR4iGCT5EBTm4D4fzLLg4LcAC9p)

### CODE REVIEW

**Comment by preschian** [comment]
### Changes:

- Heavily removed unused parts of the code in `useNft.ts`:
  - Removed `resources`, as it's only used for RMRK-related functionality
  - Since we're focusing on genart, removed the audio icon composable. We'll add it back in another PR if needed
  - @hassnian raised a good point in #10874. In this PR, that part of the code has been simplified
- Heavily removed unused types and functions
- Removed `nameWithIndex/displaySn` because it showed inconsistent results between on-chain data and our app
- Added nft store to pass data between components instead of props drilling
- Added "refresh metadata" button to refresh the metadata of the nft
- Removed RMRK-related parts (`resources/parents`)
- Added 'oda-worker' service to fetch metadata from on-chain data. Now, the gallery item page will always get metadata from `oda-worker`. First, data will be returned from the indexer, then revalidated with data from `oda-worker`

**Comment by hassnian** [comment]
CleanShot 2024-09-07 at 11 35 30@2x


should fix it https://github.com/kodadot/runners/pull/10

**Comment by preschian** [comment]
> should fix it [kodadot/runners#10](https://github.com/kodadot/runners/pull/10)

thank you

File: components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue [hunk 1]
```diff
@@ -151,4 +156,11 @@ const unlist = () => {
     errorMessage: $i18n.t('transaction.unlist.error') as string,
   })
 }
+
+const refreshMetadata = async () => {
+  if (props.collectionId && props.nftSn) {
+    await refreshOdaTokenMetadata(urlPrefix.value, props.collectionId!, props.nftSn!)
```
Comment (hassnian) [inline components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue:162]: maybe show a toast `refreshing...`, atm it looks like the click did nothing untll it refreshes

**Review by hassnian — COMMENTED** [review]
nice clean-up!

my only concern is that now the explore page does not show the sn, users wont be able differentiate nfts by sn,  

CleanShot 2024-09-07 at 13 26 36@2x

**Comment by preschian** [comment]
> my only concern is that now the explore page does not show the sn

This is from the indexer. Currently, our production/canary also returns the same results. In the past, there was a task to group similar NFTs on the explorer tab, such as "Buy The Dip" and "Wokelympics" items. I suspect it returns the collection name instead of the NFT name.

Screenshot 2024-09-07 at 16 03 26

> users wont be able differentiate nfts by sn

Actually, this is my concern about why I removed the nameWithIndex functionality. It will return inconsistent results in the token name. For example, this Monkey Business item returns `#88` on the explorer, but if a user clicks on it, the metadata name returns `#3`, not `#88`.



I've been considering fetching from `oda-worker` on the explorer to get the correct metadata. I think I will add that for AssethHub only (because the EVM indexer should be fine). The trade-off is that users will be making more requests. However, since `oda-worker` uses caching, it should be fine. What do you think?

File: utils/nft.ts [hunk 2]
```diff
@@ -4,30 +4,6 @@ import type { NFTWithMetadata } from '@/composables/useNft'
...
-    return name
-  }
-
-  // display sn if nftId is less than 4 digits
-  if (sn.length >= 4) {
-    return name
-  }
-
-  return suffixRegex.test(name) ? name : `${name} #${trimLeadingZeros(sn)}`
-}
```
Comment (Jarsen136) [inline utils/nft.ts:33]: why? There was an issue before that it wants to show the `#sn` number on the name of nft.

File: services/oda.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    image: string
+    description: string
+    generative_uri?: string
+    banner?: string
+  }
+  supply: string
+  claimed: string
+}
+
+export const queryClient = new QueryClient()
```
Comment (Jarsen136) [inline services/oda.ts:19]:
```suggestion
```

I can't see it's used somewhere.

File: services/oda.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import type { Prefix } from '@kodadot1/static'
+import { QueryClient } from '@tanstack/query-core'
```
Comment (Jarsen136) [inline services/oda.ts:2]:
```suggestion
```

File: utils/nft.ts (hunk 2 above)
Comment (preschian) [reply utils/nft.ts:33]: Refer to my comment here: [PR #10934 Comment](https://github.com/kodadot/nft-gallery/pull/10934#issuecomment-2335128843)

Just return it as it is from token metadata. I guess that one is for old drops collection, isn't it? If yes, we can update old drops metadata instead. Something like this: [Issue #217](https://github.com/kodadot/private-workers/issues/217)

Because the indexer doesn't return the proper name in some tokens, AssethHub needs an additional request. Here are the changes: [`c2fb564` (#10934)](https://github.com/kodadot/nft-gallery/pull/10934/commits/c2fb564743dec98c1a720b8f7e74903b83e48926)

File: components/gallery/GalleryItem.vue [hunk 5]
```diff
@@ -249,16 +202,14 @@ const galleryDescriptionRef = ref<{ isLewd: boolean } | null>(null)
...
-  nftMimeType,
-  nftResources,
-} = galleryItem
+const nftStore = useNftStore()
+const nft = computed(() => nftStore.nft)
+const nftMetadata = computed(() => nftStore.nftMetadata)
+const nftImage = computed(() => nftStore.nftImage)
+const nftAnimation = computed(() => nftStore.nftAnimation)
+const nftAnimationMimeType = computed(() => nftStore.nftAnimationMimeType)
+const nftMimeType = computed(() => nftStore.nftMimeType)
```
[the remark this answers is not in the document]
Comment (preschian) [reply components/gallery/GalleryItem.vue]: updated, related store getters

File: components/gallery/GalleryItemTabsPanel/GalleryItemTabsPanel.vue [hunk 6]
```diff
@@ -31,22 +31,21 @@
 
 <script setup lang="ts">
 import { NeoTabItem, NeoTabs } from '@kodadot1/brick'
-
-import type { GalleryItem } from '../useGalleryItem'
 import GalleryItemActivity from './GalleryItemActivity.vue'
 import GalleryItemChart from './GalleryItemChart.vue'
+import { useNftStore } from '@/stores/nft'
```
[the remark this answers is not in the document]
Comment (preschian) [reply components/gallery/GalleryItemTabsPanel/GalleryItemTabsPanel.vue]: updated

File: components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue (hunk 1 above)
Comment (preschian) [reply components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue:163]: added

**Comment by vikiival** [comment]
> This is from the indexer. Currently, our production/canary also returns the same results.

Well you can always open an issue in the indexer, but item metada has following lifecycle

1. At create, it will obtain metadata from collection
2. At setmetadata it is overwritten by anything 

If the metadata comes from the ipfs, they could be reuploaded again. 
If the metadata comes from our worker it also can be reindexed/reuploaded. 

I am just telling if you know there is a root issue, why not to fix it?

**Comment by preschian** [comment]
> I am just telling if you know there is a root issue, why not to fix it?

Well, I don't know which section exactly on the indexer side is responsible for this. Additionally, development there is slow because it probably needs to index so many blocks. Reference: [Stick Issue #266](https://github.com/kodadot/stick/issues/266)

So, with this fix, it won't be blocked from the indexer side.

**Comment by vikiival** [comment]
> Well, I don't know which section exactly on the indexer side is responsible for this.

> 1. At create, it will obtain metadata from collection

 - https://github.com/kodadot/stick/blob/6e00bd5614b19080c8c445b71ecac81fd7e60be1/src/mappings/nfts/mint.ts#L66
 
 > 2. At setmetadata it is overwritten by anything
 
 - https://github.com/kodadot/stick/blob/6e00bd5614b19080c8c445b71ecac81fd7e60be1/src/mappings/nfts/setMetadata.ts#L46

**Comment by preschian** [comment]
> > Well, I don't know which section exactly on the indexer side is responsible for this.
> 
> > 1. At create, it will obtain metadata from collection
> 
> * [kodadot/stick@`6e00bd5`/src/mappings/nfts/mint.ts#L66](https://github.com/kodadot/stick/blob/6e00bd5614b19080c8c445b71ecac81fd7e60be1/src/mappings/nfts/mint.ts#L66)
> 
> > 2. At setmetadata it is overwritten by anything
> 
> * [kodadot/stick@`6e00bd5`/src/mappings/nfts/setMetadata.ts#L46](https://github.com/kodadot/stick/blob/6e00bd5614b19080c8c445b71ecac81fd7e60be1/src/mappings/nfts/setMetadata.ts#L46)

can anyone try fix it on indexer side maybe? cc @kodadot/internal-dev 
for me, this is a huge pain point to develop stick on local: https://github.com/kodadot/stick/issues/266

there is two issue
1. inconsistent metadata from indexer, code in the link above
2. nft name from token-entity for explorer page

**Comment by preschian** [comment]
additional context, I already tried to fix it on indexer side: https://github.com/kodadot/stick/pull/265/files. but the PR is closed. maybe anyone have a better idea about it

**Comment by vikiival** [comment]
> can anyone try fix it on indexer side maybe? cc @kodadot/internal-dev for me, this is a huge pain point to develop stick on local: [kodadot/stick#266](https://github.com/kodadot/stick/issues/266)

- https://github.com/kodadot/stick/pull/301

> 1. inconsistent metadata from indexer, code in the link above
> 2. inconsistent nft name from token-entity for explorer page

**Review by hassnian — APPROVED** [review]

**Review by Jarsen136 — APPROVED** [review]

**Comment by vikiival** [comment]
**Copilot is generating a summary...**

This pull request includes several changes to the `components` directory, focusing on simplifying the codebase and improving the user interface. The most important changes include removing unused imports and interfaces, simplifying the `GalleryItem` component, and updating the display of NFT names.

### Codebase Simplification:

* [`components/base/types.ts`](diffhunk://#diff-3c83e231b762ebb64234971768d550dc5eae62c639d2285cf7bca9d0da5a921fL2): Removed the unused `ItemResources` import and interface. (`[[1]](diffhunk://#diff-3c83e231b762ebb64234971768d550dc5eae62c639d2285cf7bca9d0da5a921fL2)`, `[[2]](diffhunk://#diff-3c83e231b762ebb64234971768d550dc5eae62c639d2285cf7bca9d0da5a921fL21-R20)`)
* [`components/gallery/GalleryItem.vue`](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L26-L56): Simplified the `GalleryItem` component by removing unnecessary properties and methods, and refactoring to use `storeToRefs`. (`[[1]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L26-L56)`, `[[2]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L114-R84)`, `[[3]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L186-R157)`, `[[4]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L212-R191)`, `[[5]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L256-R211)`, `[[6]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L282-L287)`, `[[7]](diffhunk://#diff-5937c7434b4c50a2f0bfa46e13f06ccb54f226bf0820f7ab958781d1b0d44b02L322-L325)`)

### User Interface Improvements:

* [`components/common/CartItemDetails.vue`](diffhunk://#diff-fecbfc9681176788ffe09a2c957d6d851279bf76be90901346021e6e3f606721L31-R31): Simplified the display of NFT names by removing the `nameWithIndex` utility function. (`[components/common/CartItemDetails.vueL31-R31](diffhunk://#diff-fecbfc9681176788ffe09a2c957d6d851279bf76be90901346021e6e3f606721L31-R31)`)
* [`components/common/confirmPurchaseModal/ConfirmPurchaseItemRow.vue`](diffhunk://#diff-41e6a55b9b194a5115120fa02c746cbae6b1a222c92c3d27d19a558b986d691eL18-R18): Updated the display of NFT names to remove the serial number. (`[[1]](diffhunk://#diff-41e6a55b9b194a5115120fa02c746cbae6b1a222c92c3d27d19a558b986d691eL18-R18)`, `[[2]](diffhunk://#diff-41e6a55b9b194a5115120fa02c746cbae6b1a222c92c3d27d19a558b986d691eL37)`)
* [`components/common/shoppingCart/ShoppingCartItemRow.vue`](diffhunk://#diff-aa2b4be6571485b9ba7ff62bae4d2c8bc834df3f1904af78d1c898fd0a97b3cfL26-R26): Updated the display of NFT names to remove the serial number. (`[[1]](diffhunk://#diff-aa2b4be6571485b9ba7ff62bae4d2c8bc834df3f1904af78d1c898fd0a97b3cfL26-R26)`, `[[2]](diffhunk://#diff-aa2b4be6571485b9ba7ff62bae4d2c8bc834df3f1904af78d1c898fd0a97b3cfL63-R63)`)
* [`components/common/shoppingCart/utils.ts`](diffhunk://#diff-d322ce38e40356aee1a378d0a6bc8e96f30550408000564cf48f8d8c93cd2e7fL8): Removed the `nameWithIndex` utility function and updated related methods to display NFT names without the serial number. (`[[1]](diffhunk://#diff-d322ce38e40356aee1a378d0a6bc8e96f30550408000564cf48f8d8c93cd2e7fL8)`, `[[2]](diffhunk://#diff-d322ce38e40356aee1a378d0a6bc8e96f30550408000564cf48f8d8c93cd2e7fL52-R51)`, `[[3]](diffhunk://#diff-d322ce38e40356aee1a378d0a6bc8e96f30550408000564cf48f8d8c93cd2e7fL76-R75)`, `[[4]](diffhunk://#diff-d322ce38e40356aee1a378d0a6bc8e96f30550408000564cf48f8d8c93cd2e7fL97-R96)`, `[[5]](diffhunk://#diff-d322ce38e40356aee1a378d0a6bc8e96f30550408000564cf48f8d8c93cd2e7fL115-R114)`)

### Carousel and Media Updates:

* [`components/carousel/module/CarouselMedia.vue`](diffhunk://#diff-f6b0eb17e539188ac7005e8c1572aed9383a648ae29c460254adf1c77a78a421L7-L12): Removed the card icon and related properties from the carousel media component. (`[[1]](diffhunk://#diff-f6b0eb17e539188ac7005e8c1572aed9383a648ae29c460254adf1c77a78a421L7-L12)`, `[[2]](diffhunk://#diff-f6b0eb17e539188ac7005e8c1572aed9383a648ae29c460254adf1c77a78a421L42-L52)`)
* [`components/gallery/GalleryItemDescription.vue`](diffhunk://#diff-b3ac2906714600748232d7586158cd28f0d26773c09e0fad2d7f0a1a89174d63L155-R155): Removed the parent tab from the gallery item description component. (`[[1]](diffhunk://#diff-b3ac2906714600748232d7586158cd28f0d26773c09e0fad2d7f0a1a89174d63L155-R155)`, `[[2]](diffhunk://#diff-b3ac2906714600748232d7586158cd28f0d26773c09e0fad2d7f0a1a89174d63L226-L253)`)

### Metadata Refresh:

* [`components/gallery/GalleryItemButton/GalleryItemMoreActionBtn.vue`](diffhunk://#diff-94c0e06c26ea52817bc43c393fe0dc84204962488b2f105d0f71957fd564e419R59-R62): Added a new option to refresh metadata in the gallery item more action button. (`[[1]](diffhunk://#diff-94c0e06c26ea52817bc43c393fe0dc84204962488b2f105d0f71957fd564e419R59-R62)`, `[[2]](diffhunk://#diff-94c0e06c26ea52817bc43c393fe0dc84204962488b2f105d0f71957fd564e419R74)`, `[[3]](diffhunk://#diff-94c0e06c26ea52817bc43c393fe0dc84204962488b2f105d0f71957fd564e419R176-R182)`)

**Review by Jarsen136 — APPROVED** [review]

**Review by hassnian — APPROVED** [review]

**Comment by vikiival** [comment]
@preschian any drawbacks from this solution?

**Comment by preschian** [comment]
> @preschian any drawbacks from this solution?

the drawbacks is here: https://github.com/kodadot/nft-gallery/pull/10934#issuecomment-2335128843. On AssethHub explorer, it will fetch more additional requests because currently, the explorer doesn't show serial numbers for AssethHub.

before: https://koda.art/ahp/explore/items?search=patience&listed=false
after: https://deploy-preview-10934--koda-canary.netlify.app/ahp/explore/items?search=patience&listed=false

Once fixed from the indexer side, it will be safe to remove these lines:
https://github.com/kodadot/nft-gallery/blob/4a1b1bf3c1ef84939a3a7c9b17375c833217f8fe/components/items/ItemsGrid/ItemsGridImageTokenEntity.vue#L236-L250

**Comment by vikiival** [comment]
> Once fixed from the indexer side,

What exactly should be fixed ?

**Comment by preschian** [comment]
> > Once fixed from the indexer side,
> 
> What exactly should be fixed ?

oh, it's already reported before, refs:
- https://github.com/kodadot/nft-gallery/issues/8011
- https://github.com/kodadot/stick/issues/144
- https://github.com/kodadot/stick/pull/145/files#diff-8fde7acc9906168794ac255bb22725986fd239fb7c12760d7daf4c2e51b259fbR23-R27

since it's from indexer side, lets merge it

**Comment by vikiival** [comment]
> oh, it's already reported before

can you please open issue in the stick repo with proper description ?

**Comment by preschian** [comment]
> > oh, it's already reported before
> 
> can you please open issue in the stick repo with proper description ?

https://github.com/kodadot/stick/issues/144#issuecomment-2391166631
